### ISSUE TITLE
Rocket API flagging? (unsafe to use Necrobot)

### ISSUE THREAD

**Issue by CreativiTimothy** [NoxxDev/NecroBot#867]
Sorry if this is spam. I just feel that it is very important. In the reddit pokemongodev sub, they mentioned that the rocket API (which Necrobot is using) is not flag proof, and that Niantic can easily flag anyone using programs that use the rocket API and later ban them in a banwave. So, we're all likely to get all banned sooner or later as our accounts have probably been already flagged.

Some other third party tools use through MITM method to avoid having to use API calls, and others use one with unknown6 signature. I don't know anything about APIs, but does Necrobot already have unknown6 signature encryption or use proper signatures for API calls? If not, could we make it a number 1 priority to implement it? If we don't, all the coding for other features and bug fixes would be pointless if we're going to get flagged and banned in the 3rd banwave (probably after the Pokemon buddy update). There appears to be a pull request on the rocket API that implements unknown6 signature encryption, but the author Feroxrev seems to be afk as the pull request was 27 days ago and it still wasn't implemented. Here's the link: https://github.com/FeroxRev/Pokemon-Go-Rocket-API/pull/128

I don't know if this helps, but here are some other links related to API:
https://github.com/cyraxx/pogobuf
https://github.com/cyraxx/node-pogo-protos
https://github.com/AeonLucid/POGOProtos

**Comment by Silph-Road** [comment]
Now this is an A+ post!!  :100: 

FeroxRev has been inactive for about a month now, last commit Aug 2, 2016, Sadly, I don't think there will be any update on that repo. 

I say we invite @Gabrola, to perhaps further develop the [NecroBot-Rocket-API code](https://github.com/NoxxDev/NecroBot-Rocket-API) currently being used in Necrobot.

Related PogoDevOrg GitHub (constantly updated)
https://github.com/pogodevorg
https://github.com/pogodevorg/pgoapi-go

**Comment by dbw-holic** [comment]
What people have been doing here (humanised throws, random delays) are most likely a placebo.

**Comment by jjskuld** [comment]
Yes, unfortunately, Rocket API is kind of a mess and FeroxRev has abandoned the project.  I think Rocket API is broken in a lot of ways after looking at it.  I've got a private fork of Rocket API that better emulates the 0.35.0 client. I basically ported over parts of https://github.com/Grover-c13/PokeGOAPI-Java API to .NET.  

This would only be a band-aid since after looking at Rocket API,  I think it's so broken that it a complete re-write might be needed to better emulate a real client.

I've been considering a few options:
1) Patch Rocket API and continue to fix this base.
2) Run a Java to C# converter, manually fix the code that didn't port over. This is a big effort initially.
3) Run a Java to .NET bridge (like JNBridge) so that we could use Grover-C13's API with NecroBot.
4) Wrap Grover-C13 in a web services interface call into it from .NET using web services.

Somehow interfacing with Grover-c13's code is appealing for several reasons:
1) It's actively maintained and they focus purely on the API.
2) Grover-c13 also contributes back to AeonLucid/POGOProtos regularly so I feel comfortable that they are "experts" in this area of better reverse engineering.
3) Their inventory management is way better than Rocket API.
4) They have gym attack code working (or semi-working).
5) Preliminary captcha code checks in place.  At least they know when the server sends back a captcha challenge.

**Comment by Silph-Road** [comment]
I agree, Grover-c13's code is constantly updated.

He works on it all the time and is also very helpful as he helped me configure device settings the auth.json for iOS devices and so far so good.

You are correct, the entire PogoDevTeam (including Grover-c13 & AeonLucid)  is very focused on API and constantly analyze MITM data and improve the code.

I think interfacing with Grover-c13's code would be best for the bot. :+1: 

If the NecroBot devs need more financial support, i highly suggest updating/creating a new necro discord with the PayPal donate link. There are a lot of users who will support them. Plus it will help users get troubleshooting assistance.

Overall, this is a great project and the Devs work is much appreciated!!!! Necrobot is the best bot when it comes to features and dev dedication.

Thank you all :
@NoxxDev 
@jjskuld 
@mo0ojava 
@samuraitruong

**Comment by Gabrola** [comment]
@Silph-Road 
Thanks for inviting me on here. While I'm not currently active in PoGo dev, I'd like to chime in and drop some comments. 

Firstly, I totally agree that Rocket-API is definitely broken on all of it's current forks. The thing is that Niantic can easily flag API's claiming to be a certain version (through one of the changing unknown fields like unknown25) of the client while it doesn't correctly emulate the protocol correctly. For example I've noticed that on your fork of Rocket-API you are not adding the unknown6 signature to the initial request envelope using AuthInfo instead of AuthTicket as the seed. I think this problem affects all API's not just Rocket-API, so for example if you update the API with a change that occurred in the latest client version but your unknown25 is still pointing to an older version, you can be easily flagged. 

I think the best solution would be to focus on a specific client version and constantly analyze it's data to mimic it as close as possible. If possible it would be ideal if you can create a solution that would be able to dissect MITM requests and try to reconstruct them using the API and verifying if the requests match the originals.

As for interfacing with the Java API, I do also agree that would be the best course of action. I think the best option would be to create a RESTful API wrapper for that API and use that. That way you'll benefit from the constant updates of their API without having much or any overhead of updating a parallel codebase in .NET as well as contributing any improvements to their code.

**Comment by vfuyong** [comment]
Reusability is the champion!
